metric/metricexport/reader.go

```diff
@@ -0,0 +1,175 @@
+// Copyright 2019, OpenCensus Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package metricexport
+
+import (
+	"fmt"
+	"time"
+
+	"context"
+	"go.opencensus.io/metric/metricdata"
+	"go.opencensus.io/metric/metricproducer"
+	"go.opencensus.io/trace"
+	"sync"
+)
+
+var (
+	defaultSampler             = trace.ProbabilitySampler(0.0001)
+	errReportingIntervalTooLow = fmt.Errorf("reporting interval less than %d", MinimumReportingDuration)
+	errAlreadyStarted          = fmt.Errorf("already started")
+	errIntervalReaderNil       = fmt.Errorf("interval reader is nil")
+	errExporterNil             = fmt.Errorf("exporter is nil")
+	errReaderNil               = fmt.Errorf("reader is nil")
+)
+
+// IntervalReader periodically reads metrics from all producers registered
+// with producer manager and exports those metrics using provided
+// exporter. Call Reader.Stop() to stop the reader.
+type IntervalReader struct {
+	// ReportingInterval it the time duration between two consecutive
+	// metrics reporting. DefaultReportingDuration  is used if it is not set.
+	// It cannot be set lower than MinimumReportingDuration.
+	ReportingInterval time.Duration
+
+	exporter   Exporter
+	timer      *time.Ticker
+	quit, done chan bool
+	mu         sync.RWMutex
+	reader     *Reader
+}
+
+// Reader reads metrics from all producers registered
+// with producer manager and exports those metrics using provided
+// exporter.
+type Reader struct {
```

Comment:
Probably for this we should do a WithOption pattern because does not have a start method.